http/zabbix_http.go

```diff
@@ -152,522 +115,696 @@ func hostCreate(nodes map[string]interface{}, rw http.ResponseWriter) {
+ */
+func apiAlert(rw http.ResponseWriter, req *http.Request) {
+}
+
+/**
+ * @function name:	func apiParser(rw http.ResponseWriter, req *http.Request)
+ * @description:	This function parses the method of API request.
+ * @related issues:	OWL-085
+ * @param:			rw http.ResponseWriter
+ * @param:			req *http.Request
+ * @return:			void
+ * @author:			Don Hsieh
+ * @since:			09/11/2015
+ * @last modified: 	09/23/2015
+ * @called by:		http.HandleFunc("/api", apiParser)
+ *					 in func main()
+ */
 func apiParser(rw http.ResponseWriter, req *http.Request) {
+	log.Println("func apiParser(rw http.ResponseWriter, req *http.Request)")
 	buf := new(bytes.Buffer)
 	buf.ReadFrom(req.Body)
-	s := buf.String() // Does a complete copy of the bytes in the buffer.
-
-	json, err := simplejson.NewJson(buf.Bytes())
-	if err != nil {
-		log.Println(err.Error())
-	}
-
-	f, err := os.OpenFile("falcon_api.log", os.O_RDWR|os.O_CREATE|os.O_APPEND, 0666)
-	if err != nil {
-		log.Fatal("error opening file: %v", err)
-	}
-	defer f.Close()
-	log.SetOutput(f)
-
-	var nodes = make(map[string]interface{})
-	nodes, _ = json.Map()
-
-	method := nodes["method"]
-	delete(nodes, "method")
-	delete(nodes, "auth")
-
-	if method == "host.create" {
-		hostCreate(nodes, rw)
-	} else if method == "host.delete" {
-		hostDelete(nodes, rw)
-	} else if method == "host.update" {
-		hostUpdate(nodes, rw)
-	} else if method == "host.exists" {
-		// hostExist(params)
-	} else if method == "hostgroup.create" {
-		hostgroupCreate(nodes, rw)
-	} else if method == "hostgroup.delete" {
-		hostgroupDelete(nodes, rw)
-	} else if method == "hostgroup.update" {
-		hostgroupUpdate(nodes, rw)
-	} else if method == "hostgroup.exists" {
-		// hostgroupExist(params)
-	} else if method == "template.create" {
-		templateCreate(nodes, rw)
-	} else if method == "template.delete" {
-		templateDelete(nodes, rw)
-	} else if method == "template.update" {
-		templateUpdate(nodes, rw)
-	} else if method == "template.exists" {
-		// templateExist(params)
-	}
-}
-
-func parseConfig(cfg string) {
-	if !file.IsExist(cfg) {
-		log.Fatalln("config file:", cfg, "is not existent. maybe you need `mv cfg.example.json cfg.json`")
-	}
-	ConfigFile = cfg
-	configContent, err := file.ToTrimString(cfg)
-	if err != nil {
-		log.Fatalln("read config file:", cfg, "fail:", err)
-	}
+	log.Println(buf.Len())
+	if buf.Len() == 0 {
+		apiAlert(rw, req)
+	} else {
+		s := buf.String() // Does a complete copy of the bytes in the buffer.
+		log.Println("s =", s)
+		json, err := simplejson.NewJson(buf.Bytes())
+		if err != nil {
+			log.Println(err.Error())
+		}
 
-	var c GlobalConfig
-	err = json.Unmarshal([]byte(configContent), &c)
-	if err != nil {
-		log.Fatalln("parse config file:", cfg, "fail:", err)
-		return
+		var nodes = make(map[string]interface{})
+		nodes, _ = json.Map()
+
+		method := nodes["method"]
+		log.Println(method)
+		delete(nodes, "method")
+		delete(nodes, "auth")
+
+		if method == "host.create" {
```

Comment:
Considering the testing issues:
    1.   **if-else structure** can be replaced by **factory pattern**. 
    2.  **hostCreate, hostDelete, hostUpdate... functions**  can be replaced by **several objects which implement the same interface**. **Factory** returns an the object, and the **function apiParser** uses the interface of the object to process the output. 
    3.  Input arguments and output arguments can be separated in **function hostCreate** instead of all stay in just one argument **nodes**.
  
Then, the logic of **function apiParser** will be further simplified and **unit test** will be written possibly.  
Ref:
http://oomusou.io/tdd/tdd-di/